### DISCUSSION TITLE
Create LUT to map input sample data to target

### DISCUSSION THREAD

**Discussion by jfitch12** [colour-science/colour#1299] [Q&A]
Is it possible to create a 3D LUT with `colour` package or `OCIO` package that accurately maps input sample data to the desired output target? 

Example:
I have a table of 9x9x9 measurements from a projector and I want to create a 9x9x9 LUT that corrects the signal so that the projector displays an accurate image e.g. Rec 709 Gamma 2.4.

I tried using a color space transform, but that does not correct well the uniformity issues in the sample data. I've also looked into `colour.colour_correction` but that seems also to produce a 3x3 matrix.

Appreciate any sage advice. 😄 

Cheers!

J

**Comment by tjdcs** [comment]
You need to use a color model for the projector to estimate how far off each color measurement is in terms of RGB values. THen adjust the values in the LUT by the inverse amount.

**Comment by tjdcs** [comment]
Do you have any example data or existing code that we can help review?

**Comment by jfitch12** [comment]
@tjdcs 
Hi Tucker,
thanks for replying!

I just measured a fresh new 10x10x10 on a desktop display in front of me to test out my results.
Here are the main color points from that profile, let me know if you want the entire data set - I'm not sure if I can attach a file here and I do not want to flood the forum with larger data sets.

Data:
```
the target destination: Rec 709 Gamma 2.2
10x10x10 grid stimulus steps measured: [0, 0.111111, 0.222222, 0.333333, 0.444444, 0.555556, 0.666667, 0.777778, 0.888889, 1]
XYZ raw result for W 1,1,1: [ 180.86004610217705, 170.23242506811988, 190.45602355533302 ]
XYZ raw result for R 1,0,0: [ 112.04687701223439, 52.28141689373297, 0.9602210314836214 ]
XYZ raw result for G 0,1,0: [ 38.9051822803817, 106.91336971350614, 17.606401666424134 ]
XYZ raw result for B 0,0,1: [ 31.108667095943336, 12.812125340599454, 174.1869838242821 ]
XYZ raw result 0.555556,0.777778,0.888889: [ 65.99804123711341, 73.92491087014169, 127.2108258054445 ]
```
XYZ-to-RGB for display's color space:
```
[
  [ 3.240969941904523, -1.5373831775700941, -0.4986107602930035 ],
  [ -0.9692436362808787, 1.875967501507719, 0.041555057407175564 ],
  [ 0.05563007969699363, -0.20397695888897668, 1.0569715142428784 ]
]
```

RGB-to-RGB from display's color space to destination Rec 709
```
[
  [ 0.6111533676428355, 0.17582546428238588, 0.018081285469724206 ],
  [ 0.039562781035707845, 1.0667738822694046, -0.005229949471805008 ],
  [ 0.011729550324575165, 0.009346738934438336, 0.9402171837427313 ]
]
```

my current process:
1 - normalize all XYZ raw by White Y nits (170.23242506811988)
2 - create a XYZ-to-RGB matrix from the display's primaries R,G,B, see above
3 - map a data sample point into the display's RGB linear space using the XYZ-to-RGB matrix
4 - take the RGB stimulus that was used to measure, and reverse the gamma encoding of target 2.2 to result in linear RGB space
5 - with both the reference and the result in same linear RGB space, calculate the rgb delta from reference to result and add the delta to reference rgb value
    >>> is this the same as "adjusting by the inverse amount" ?
6 - map the new rgb value to the target rec 709 color space using the RGB-to-RGB matrix, see above
7 - clip negative values
8 - encode with target/destination Gamma of 2.2
9 - clip 0-1, store in 10x10x10 LUT

If I do this for the 5th measurement sample above for stimulus 0.555556,0.777778,0.888889 the final RGB LUT result is:
```
[ 0.605838599254416, 0.853706120049507, 0.9165080283712354 ]
```
this result, while closer, still measures too far off.
Let me know if you need anything else.

Thanks a bunch!
John

**Comment by jfitch12** [comment]
@tjdcs 

> You need to use a color model for the projector to estimate how far off each color measurement is in terms of RGB values. THen adjust the values in the LUT by the inverse amount.

Could you elaborate in detail on how to do this?
I suspect this may be one of my mistakes.

And here is the full profile attached.
[Dell_10x10x10.csv](https://github.com/user-attachments/files/17322477/Dell_10x10x10.csv)

Thx!
